src/AddressSpace.cc

```diff
@@ -881,22 +884,33 @@ void AddressSpace::unmap_internal(remote_ptr<void> addr, ssize_t num_bytes) {
 
     Mapping m = move(mm);
     mem.erase(m.map);
+
+    // Also unmap any corresponding mapping from the local address space
+    if (m.local_addr) {
+      uint8_t* start_addr =
+          m.local_addr + (max(m.map.start(), rem.start()) - m.map.start());
+      uint8_t* stop_addr =
+          m.local_addr + (min(m.map.end(), rem.end()) - m.map.end());
+      munmap(start_addr, stop_addr - start_addr);
+    }
```

Comment:
In particular this would interact with `remap`; we `unmap_internal` the remapped memory. So either the caller already remapped the local mapping, in which case we shouldn't try to unmap it ourselves (e.g. consider a racing thread that mmaps the same area), or the caller didn't remap the local mapping in which case we'll be destroying it and leaving an invalid `local_addr` pointer.